File: packages/jetstream_wireformat/src/primitives.ts
/**
 * r[impl jetstream.wireformat.ts.interface]
 * r[impl jetstream.wireformat.trait]
 */

import { BinaryReader } from './binary-reader.js';
import { BinaryWriter } from './binary-writer.js';

/**
 * WireFormat interface - every type that can be transmitted over the wire
 * MUST implement this interface providing three operations.
 */
export interface WireFormat<T> {
  byteSize(value: T): number;
  encode(value: T, writer: BinaryWriter): void;
  decode(reader: BinaryReader): T;
}

/**
 * r[impl jetstream.wireformat.u8]
 * r[impl jetstream.wireformat.ts.u8]
 */
export const u8Codec: WireFormat<number> = {
  byteSize(_value: number): number {
    return 1;
  },
  encode(value: number, writer: BinaryWriter): void {
    writer.writeU8(value);
  },
  decode(reader: BinaryReader): number {
    return reader.readU8();
  },
};

/**
 * r[impl jetstream.wireformat.u16]
 * r[impl jetstream.wireformat.ts.u16]
 */
export const u16Codec: WireFormat<number> = {
  byteSize(_value: number): number {
    return 2;
  },
  encode(value: number, writer: BinaryWriter): void {
    writer.writeU16(value);
  },
  decode(reader: BinaryReader): number {
    return reader.readU16();
  },
};

/**
 * r[impl jetstream.wireformat.u32]
 * r[impl jetstream.wireformat.ts.u32]
 */
export const u32Codec: WireFormat<number> = {
  byteSize(_value: number): number {
    return 4;
  },
  encode(value: number, writer: BinaryWriter): void {
    writer.writeU32(value);
  },
  decode(reader: BinaryReader): number {
    return reader.readU32();
  },
};

/**
 * r[impl jetstream.wireformat.u64]
 * r[impl jetstream.wireformat.ts.u64]
 */
export const u64Codec: WireFormat<bigint> = {
  byteSize(_value: bigint): number {
    return 8;
  },
  encode(value: bigint, writer: BinaryWriter): void {
    writer.writeU64(value);
  },
  decode(reader: BinaryReader): bigint {
    return reader.readU64();
  },
};

/**
 * r[impl jetstream.wireformat.u128]
 * r[impl jetstream.wireformat.ts.u128]
 */
export const u128Codec: WireFormat<bigint> = {
  byteSize(_value: bigint): number {
    return 16;
  },
  encode(value: bigint, writer: BinaryWriter): void {
    const mask64 = 0xFFFFFFFFFFFFFFFFn;
    const lower = value & mask64;
    const upper = (value >> 64n) & mask64;
    writer.writeU64(lower);
    writer.writeU64(upper);
  },
  decode(reader: BinaryReader): bigint {
    const lower = reader.readU64();
    const upper = reader.readU64();
    return (upper << 64n) | lower;
  },
};

/**
 * r[impl jetstream.wireformat.i16]
 * r[impl jetstream.wireformat.ts.i16]
 */
export const i16Codec: WireFormat<number> = {
  byteSize(_value: number): number {
    return 2;
  },
  encode(value: number, writer: BinaryWriter): void {
    writer.writeI16(value);
  },
  decode(reader: BinaryReader): number {
    return reader.readI16();
  },
};

/**
 * r[impl jetstream.wireformat.i32]
 * r[impl jetstream.wireformat.ts.i32]
 */
export const i32Codec: WireFormat<number> = {
  byteSize(_value: number): number {
    return 4;
  },
  encode(value: number, writer: BinaryWriter): void {
    writer.writeI32(value);
  },
  decode(reader: BinaryReader): number {
    return reader.readI32();
  },
};

/**
 * r[impl jetstream.wireformat.i64]
 * r[impl jetstream.wireformat.ts.i64]
 */
export const i64Codec: WireFormat<bigint> = {
  byteSize(_value: bigint): number {
    return 8;
  },
  encode(value: bigint, writer: BinaryWriter): void {
    writer.writeI64(value);
  },
  decode(reader: BinaryReader): bigint {
    return reader.readI64();
  },
};

/**
 * r[impl jetstream.wireformat.i128]
 * r[impl jetstream.wireformat.ts.i128]
 */
export const i128Codec: WireFormat<bigint> = {
  byteSize(_value: bigint): number {
    return 16;
  },
  encode(value: bigint, writer: BinaryWriter): void {
    const mask64 = 0xFFFFFFFFFFFFFFFFn;
    // Lower 64 bits as unsigned
    const lower = value & mask64;
    writer.writeU64(lower);
    // Upper 64 bits as signed
    const upper = value >> 64n;
    writer.writeI64(upper);
  },
  decode(reader: BinaryReader): bigint {
    // Lower 64 bits as unsigned
    const lower = reader.readU64();
    // Upper 64 bits as signed
    const upper = reader.readI64();
    return (upper << 64n) | (lower & 0xFFFFFFFFFFFFFFFFn);
  },
};

/**
 * r[impl jetstream.wireformat.f32]
 * r[impl jetstream.wireformat.ts.f32]
 */
export const f32Codec: WireFormat<number> = {
  byteSize(_value: number): number {
    return 4;
  },
  encode(value: number, writer: BinaryWriter): void {
    writer.writeF32(value);
  },
  decode(reader: BinaryReader): number {
    return reader.readF32();
  },
};

/**
 * r[impl jetstream.wireformat.f64]
 * r[impl jetstream.wireformat.ts.f64]
 */
export const f64Codec: WireFormat<number> = {
  byteSize(_value: number): number {
    return 8;
  },
  encode(value: number, writer: BinaryWriter): void {
    writer.writeF64(value);
  },
  decode(reader: BinaryReader): number {
    return reader.readF64();
  },
};

/**
 * r[impl jetstream.wireformat.bool]
 * r[impl jetstream.wireformat.ts.bool]
 */
export const boolCodec: WireFormat<boolean> = {
  byteSize(_value: boolean): number {
    return 1;
  },
  encode(value: boolean, writer: BinaryWriter): void {
    writer.writeU8(value ? 0x01 : 0x00);
  },
  decode(reader: BinaryReader): boolean {
    const byte = reader.readU8();
    if (byte === 0) return false;
    if (byte === 1) return true;
    throw new Error(`invalid byte for bool: ${byte}`);
  },
};

/**
 * r[impl jetstream.wireformat.unit]
 * r[impl jetstream.wireformat.ts.unit]
 */
export const unitCodec: WireFormat<undefined> = {
  byteSize(_value: undefined): number {
    return 0;
  },
  encode(_value: undefined, _writer: BinaryWriter): void {
    // nothing to write
  },
  decode(_reader: BinaryReader): undefined {
    return undefined;
  },
};

File: Sources/JetStreamWireFormat/Primitives.swift
// JetStream WireFormat — Primitive Types
// Copyright (c) 2024, Sevki <s@sevki.io>
// SPDX-License-Identifier: BSD-3-Clause

import Foundation

// MARK: - UInt8

// r[impl jetstream.wireformat.u8]
// r[impl jetstream.wireformat.swift.u8]
extension UInt8: WireFormat {
    public func byteSize() -> UInt32 { 1 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeU8(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> UInt8 {
        try reader.readU8()
    }
}

// MARK: - UInt16

// r[impl jetstream.wireformat.u16]
// r[impl jetstream.wireformat.swift.u16]
extension UInt16: WireFormat {
    public func byteSize() -> UInt32 { 2 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeU16(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> UInt16 {
        try reader.readU16()
    }
}

// MARK: - UInt32

// r[impl jetstream.wireformat.u32]
// r[impl jetstream.wireformat.swift.u32]
extension UInt32: WireFormat {
    public func byteSize() -> UInt32 { 4 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeU32(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> UInt32 {
        try reader.readU32()
    }
}

// MARK: - UInt64

// r[impl jetstream.wireformat.u64]
// r[impl jetstream.wireformat.swift.u64]
extension UInt64: WireFormat {
    public func byteSize() -> UInt32 { 8 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeU64(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> UInt64 {
        try reader.readU64()
    }
}

// MARK: - UInt128 (custom struct for Swift <6 compatibility)

// r[impl jetstream.wireformat.u128]
// r[impl jetstream.wireformat.swift.u128]
/// A 128-bit unsigned integer for wire format encoding.
/// Stored as (low: UInt64, high: UInt64). Encoded low-first (little-endian).
public struct WireUInt128: Equatable, Hashable {
    public var low: UInt64
    public var high: UInt64

    public init(low: UInt64, high: UInt64) {
        self.low = low
        self.high = high
    }

    public init(_ value: UInt64) {
        self.low = value
        self.high = 0
    }
}

extension WireUInt128: WireFormat {
    public func byteSize() -> UInt32 { 16 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeU64(low)
        writer.writeU64(high)
    }

    public static func decode(reader: inout BinaryReader) throws -> WireUInt128 {
        let low = try reader.readU64()
        let high = try reader.readU64()
        return WireUInt128(low: low, high: high)
    }
}

// MARK: - Int16

// r[impl jetstream.wireformat.i16]
// r[impl jetstream.wireformat.swift.i16]
extension Int16: WireFormat {
    public func byteSize() -> UInt32 { 2 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeI16(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> Int16 {
        try reader.readI16()
    }
}

// MARK: - Int32

// r[impl jetstream.wireformat.i32]
// r[impl jetstream.wireformat.swift.i32]
extension Int32: WireFormat {
    public func byteSize() -> UInt32 { 4 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeI32(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> Int32 {
        try reader.readI32()
    }
}

// MARK: - Int64

// r[impl jetstream.wireformat.i64]
// r[impl jetstream.wireformat.swift.i64]
extension Int64: WireFormat {
    public func byteSize() -> UInt32 { 8 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeI64(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> Int64 {
        try reader.readI64()
    }
}

// MARK: - Int128 (custom struct for Swift <6 compatibility)

// r[impl jetstream.wireformat.i128]
// r[impl jetstream.wireformat.swift.i128]
/// A 128-bit signed integer for wire format encoding.
/// Stored as (low: UInt64, high: Int64). Encoded low-first (little-endian).
public struct WireInt128: Equatable, Hashable {
    public var low: UInt64
    public var high: Int64

    public init(low: UInt64, high: Int64) {
        self.low = low
        self.high = high
    }

    public init(_ value: Int64) {
        self.low = UInt64(bitPattern: value)
        self.high = value < 0 ? -1 : 0
    }
}

extension WireInt128: WireFormat {
    public func byteSize() -> UInt32 { 16 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeU64(low)
        writer.writeI64(high)
    }

    public static func decode(reader: inout BinaryReader) throws -> WireInt128 {
        let low = try reader.readU64()
        let high = try reader.readI64()
        return WireInt128(low: low, high: high)
    }
}

// MARK: - Float

// r[impl jetstream.wireformat.f32]
// r[impl jetstream.wireformat.swift.f32]
extension Float: WireFormat {
    public func byteSize() -> UInt32 { 4 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeF32(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> Float {
        try reader.readF32()
    }
}

// MARK: - Double

// r[impl jetstream.wireformat.f64]
// r[impl jetstream.wireformat.swift.f64]
extension Double: WireFormat {
    public func byteSize() -> UInt32 { 8 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeF64(self)
    }

    public static func decode(reader: inout BinaryReader) throws -> Double {
        try reader.readF64()
    }
}

// MARK: - Bool

// r[impl jetstream.wireformat.bool]
// r[impl jetstream.wireformat.swift.bool]
extension Bool: WireFormat {
    public func byteSize() -> UInt32 { 1 }

    public func encode(writer: inout BinaryWriter) throws {
        writer.writeU8(self ? 1 : 0)
    }

    public static func decode(reader: inout BinaryReader) throws -> Bool {
        let byte = try reader.readU8()
        switch byte {
        case 0: return false
        case 1: return true
        default: throw WireFormatError.invalidBoolByte(byte)
        }
    }
}

// MARK: - Unit (Void equivalent)

// r[impl jetstream.wireformat.unit]
// r[impl jetstream.wireformat.swift.unit]
/// The unit type, encoding zero bytes on the wire.
public struct WireUnit: Equatable, WireFormat {
    public init() {}

    public func byteSize() -> UInt32 { 0 }

    public func encode(writer: inout BinaryWriter) throws {}

    public static func decode(reader: inout BinaryReader) throws -> WireUnit {
        WireUnit()
    }
}
